sources/optimizer/mma.f90

```
@@ -196,268 +220,115 @@ subroutine mma_init_json(this, x, n, m, json, scale, auto_scale)
-    call this%xmin%init(n)
-
-    !internal dummy variables for MMA
-    call this%p0j%init(n)
-    call this%q0j%init(n)
-    call this%pij%init(m,n)
-    call this%qij%init(m,n)
-    call this%bi%init(m)
-
-    !---nesessary for KKT check after updating df0dx, fval, dfdx --------
-    call this%y%init(m)
-    call this%lambda%init(m)
-    call this%s%init(m)
-    call this%mu%init(m)
-    call this%xsi%init(n)
-    call this%eta%init(n)
-
-    this%a0 = a0
-    this%a%x = a
-    this%c%x = c
-    this%d%x = d
-
-    !setting the bounds for the design variable based on the problem
-    this%xmax%x = xmax
-    this%xmin%x = xmin
-
-    this%low%x(:) = minval(x)
-    this%upp%x(:) = maxval(x)
-
-    !setting KKT norms to a large number for the initial design
-    this%residumax = huge(0.0_rp)
-    this%residunorm = huge(0.0_rp)
-
-    ! ------------------------------------------------------------------------ !
-    ! Assign defaults if nothing is parsed
 
-    ! based on the Cpp Code by Niels
-    if (.not. present(epsimin)) this%epsimin = 1.0e-9_rp * sqrt(real(m + n, rp))
-    if (.not. present(max_iter)) this%max_iter = 100
+    
+    ! Select backend type
+    select case (this%backnd)
+    case ("cpu")
+       call mma_init_attributes_cpu(this, x, n, m, a0, a, c, d, xmin, xmax, &
+          max_iter, epsimin, asyinit, asyincr, asydecr)
+       if (pe_rank == 0) then
+          print *, "MMA initialized with CPU backend!"
+       end if
+    case ("cuda")
+       call mma_init_attributes_device(this, x, n, m, a0, a, c, d, xmin, xmax, &
+          max_iter, epsimin, asyinit, asyincr, asydecr)
+       if (pe_rank == 0) then
+          print *, "MMA initialized with CUDA backend!"
+       end if
+    case default
+       call mma_init_attributes_cpu(this, x, n, m, a0, a, c, d, xmin, xmax, &
+          max_iter, epsimin, asyinit, asyincr, asydecr)
+       if (pe_rank == 0) then
+          print *, "Unknown backend, MMA initialized with CPU backend!"
+       end if
```

Please throw an error instead of saving the user, they will not know why
performance is not good enough.